### PR TITLE
refactor(v2): make dependency optional 

### PR DESCRIPTION
**Pull request by samsja**
# context 

this pr allow to import docarray without installing torch or protobuf

### CODE REVIEW

File: docarray/typing/__init__.py [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
-    'TorchTensor',
     'NdArray',
     'Embedding',
     'ImageUrl',
     'TextUrl',
     'AnyUrl',
     'ID',
     'Tensor',
-    'TorchEmbedding',
 ]
```
Comment (JohannesMessner) [inline docarray/typing/__init__.py:22]: Can we add the `try` for the import at the bottom, then we can do:
```python
try:
    import torch  # noqa: F401
except ImportError:
    pass
else:
    from docarray.typing.tensor import TorchEmbedding, TorchTensor  # noqa: F401
    __all__.extend(['TorchEmbedding', 'TorchTensor'])
```

File: docarray/typing/tensor/embedding/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+except ImportError:
+    pass
+else:
+    from docarray.typing.tensor.embedding.embedding_torch import TorchEmbedding  # type: ignore # noqa: [F401,E510]
+
+
+from docarray.typing.tensor.embedding.embedding import Embedding
+from docarray.typing.tensor.embedding.embedding_ndarray import NdArrayEmbedding
+
+__all__ = ['NdArrayEmbedding', 'Embedding']
```
Comment (JohannesMessner) [inline docarray/typing/tensor/embedding/__init__.py:12]: same here

File: docarray/typing/tensor/embedding/embedding.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from typing import Union
+
+from docarray.typing.tensor.embedding.embedding_ndarray import NdArrayEmbedding
+
+# from docarray.typing.tensor.embedding.embedding_torch import TorchEmbedding
+
+Embedding = Union[NdArrayEmbedding]  # , TorchEmbedding]
```
Comment (JohannesMessner) [inline docarray/typing/tensor/embedding/embedding.py:7]: What's this? If torch is installed the `Embedding` should also be able to be a `TorchEmbedding`

File: docarray/typing/tensor/tensor.py [hunk 4]
```diff
@@ -1,6 +1,8 @@
...
+from typing import TYPE_CHECKING, Union
 
 from docarray.typing.tensor.ndarray import NdArray
-from docarray.typing.tensor.torch_tensor import TorchTensor
 
-Tensor = Union[NdArray, TorchTensor]
+if TYPE_CHECKING:
+    from docarray.typing.tensor.torch_tensor import TorchTensor
+
+Tensor = Union[NdArray, 'TorchTensor']
```
Comment (JohannesMessner) [inline docarray/typing/tensor/tensor.py:8]: Does this (`'TorchTensor'` in the Union) really work? Can we test it?

**Review by JohannesMessner — CHANGES_REQUESTED** [review]

File: docarray/typing/__init__.py (hunk 1 above)
Comment (samsja) [reply docarray/typing/__init__.py:22]: > Can we add the `try` for the import at the bottom, then we can do:
> 
> ```python
> try:
>     import torch  # noqa: F401
> except ImportError:
>     pass
> else:
>     from docarray.typing.tensor import TorchEmbedding, TorchTensor  # noqa: F401
>     __all__.extend(['TorchEmbedding', 'TorchTensor'])
> ```

I had the same idea and I was wondering if this is good practice ? Probably okayish lets do it

File: docarray/typing/tensor/embedding/embedding.py (hunk 3 above)
Comment (samsja) [reply docarray/typing/tensor/embedding/embedding.py:7]: my bad I forgot this one

File: docarray/typing/tensor/tensor.py (hunk 4 above)
Comment (samsja) [reply docarray/typing/tensor/tensor.py:8]: What kind of test do you have in mind ?

**Review by JohannesMessner — APPROVED** [review]

**Pull request merged**